src/Npgsql/Util/TimeoutCancellationTokenSourceWrapper.cs

```diff
@@ -0,0 +1,66 @@
+﻿using System;
+using System.Threading;
+
+namespace Npgsql.Util
+{
+    /// <summary>
+    /// Internal struct wrapping a <see cref="CancellationTokenSource"/> for timeouts.
+    /// </summary>
+    /// <remarks>
+    /// Since there's no way to reset a <see cref="CancellationTokenSource"/> once it was cancelled,
+    /// we need to make sure that an existing cancellation token source hasn't been cancelled,
+    /// every time we start it (see https://github.com/dotnet/runtime/issues/4694).
+    /// </remarks>
+    class TimeoutCancellationTokenSourceWrapper : IDisposable
+    {
+        public TimeSpan Timeout { get; set; }
+        CancellationTokenSource _timeoutCts = new CancellationTokenSource();
+
+        public TimeoutCancellationTokenSourceWrapper(TimeSpan timeout) => Timeout = timeout;
+
+        /// <summary>
+        /// Set the timeout on the wrapped <see cref="CancellationTokenSource"/>
+        /// and make sure that it hasn't been cancelled yet
+        /// </summary>
+        public void Start()
```

Comment:
This should probably just return the CancellationToken (it may be the same, or different if we allocated a new CTS).